README.md
pytivo-video-mgr2
=================

Only slightly modified from jbernardis' original to better support my Synology NAS as well as a few personal customizations.


vidmgr/changelog.txt
PyTivo Video Manager - change log
Version 2.0c - 10/2/2011
- added logic to sort root directory - either sort physical shares and virtual shares separately or together
- fixed bug that did not allow cache to be dynamically built a second time

Version 2.0d - 10/27/2011
- added an option usefileid, default true, to turn off the file uniqueness checking which can be expensive under windows

Version 2.0e - 10/28/2011
- fixed a bug in that prevented virtual share attributes from being parsed
correctly from the vidmgr.ini file
- added the alpha virtual share type
- added the shares= option to the vidmgi.ini to limit virtual shares to a list
of pytivo shares

Version 2.0f - 11/16/2011
- changed implementation of harvesters to use subclassing
- minor changes to the subtitle when grouping is being used
- fixed bug reported by bwall23 where port number was not successfully retrieved from pytivo config file

Version 2.0g - 11/17/2011
- added __fileName and __filePath as automatic metadata items.  These are also in the default ignore list so 
they don't show up in the info view.  If you want to see them, you need to redefine metaignore (in vidmgr.ini)
to NOT include these names
- Fixed the sorting logic that caused titles like "Alien" to sort AFTER "Alien 3"
- added new vidmgr.ini option - ignorearticle - that will ignore an article (the, a, or an) at the beginning of a title.
Thus, "The Abyss" will sort with the A's instead of the T's.  The default value of this is True so this new behavior
will occur unless you change it to False.  This also affects alpha virtual shares: The Abyss will be included with the 
A's and not with the T's 

Version 2.0h - 11/23/2011
- corrected error in object constructor routine - reported on TCF by Lrhorer

Version 2.0i - 11/29/2011
- added thumbfolderfn to ini file with a default value of 'folder.jpg' - allows you to change the name of the folder file

Version 2.1 - 2/4/2012
- fixed major bug found by cweb.  There were actually 2 bugs.
    1) The cache was being built incorrectly - references by virtual shares were not being properly accounted for.  This problem
    would actually solve itself if the cache was written to disk and then read back, but if the cache was built dynamically, 
    this "correction" did not happen.  As a result, files could be deleted even though there were multiple references, which
    brings us to the second bug:
    2) The logic to see if a file was deletable was wrong.  I was attempting to prevent deletion of a file that had multiple
    links (logical or physical) just to keep things simple for myself.  Instead I prevented deletion if there were multiple references.
    Since virtual shares create additional references to the same file, this virtually precluded deleting most files.  The logic was
    fixed to be based on whether or not there are multiple file system links.
    
Version 2.1a - 1/19/2012
- added tagorder configuration parameter to specify a separate sort order for virtual share tags as opposed to videos

Version 2.1b - 4/23/2012
- fixed a bug reported by Larrs - when selecting a tivo from a list with more that 1 page's worth, pagination was not handled correctly

6/24/2012 - without renumbering release - I added the -v option to BuildCache.py to receive additional information about the cache
build process to help debug metadata issues - as requested by lrhorer.

__init__.py
from hme import *
import os
import socket
import metadata
import cPickle as pickle
import ConfigParser
import urllib
from time import asctime
from string import maketrans
from ThumbCache import ThumbCache
import Config
from Config import ( screenWidth, screenHeight, titleYPos, subTitleYPos, 
		MYKEY_PUSHCOMPLETE, MYKEY_DELETECONFIRM, MYKEY_DELETECOMPLETE, MYKEY_DELETECANCEL, MYKEY_REBUILDCACHE,
		infoWidth, infoHeight, thumbnailheight, thumbnailwidth, thumbcachesize, ConfigError )

from ListDisplayManager import ListDisplayManager
from DetailDisplayManager import DetailDisplayManager
from VideoCache import VideoCache
from VideoFile import VideoFile
from VideoDir import VideoDir
from InfoView import InfoView
from SubMenu import SubMenu
from Push import Push
from MessageBox import MessageBox

TITLE = 'PyTivo Video Manager'
version = '2.0'

print asctime(), TITLE + " version " + version + " starting"

AppPath = os.path.dirname(__file__)

tc = ThumbCache(AppPath, thumbcachesize, thumbnailwidth, thumbnailheight)

class Images:
	def __init__(self, app):
		self.Background    = self.loadimage(app, 'background')
		self.CueUp         = self.loadimage(app, 'cueup')
		self.CueDown       = self.loadimage(app, 'cuedown')
		self.CueLeft       = self.loadimage(app, 'cueleft')
		self.HiLite        = self.loadimage(app, 'hilite')
		self.SubMenu       = self.loadimage(app, 'submenu')
		self.MenuHiLite    = self.loadimage(app, 'menuhilite')
		self.MenuBkg       = self.loadimage(app, 'menubkg')
		self.IconFolder    = self.loadimage(app, 'folder')
		self.IconVideo     = self.loadimage(app, 'video')
		self.IconDVDFolder = self.loadimage(app, 'dvdfolder')
		self.IconDVDVideo  = self.loadimage(app, 'dvdvideo')
		self.PleaseWait    = self.loadimage(app, 'pleasewait')
		self.Info          = self.loadimage(app, 'info')
		self.InfoUp        = self.loadimage(app, 'infoup')
		self.InfoDown      = self.loadimage(app, 'infodown')
		self.MsgBox        = self.loadimage(app, 'msgbox')
	
	def loadimage(self, app, name):
		skin = app.opts['skin']
		if skin != None:
			fn = os.path.join(AppPath, 'skins', skin, name + ".png")
			if os.path.exists(fn):
				return Image(app, fn)
			
		fn = os.path.join(AppPath, 'skins', name + ".png")
		if os.path.exists(fn):
			return Image(app, fn)
		
		if skin == None:
			print "image '" + name + "' missing for default skin"
		else:
			print "image '" + name + "' missing for skin '" + skin + "'"
		return None
			

class Fonts:
	def __init__(self, app):
		self.fnt16 = Font(app, size=16)
		self.fnt20 = Font(app, size=20)
		self.fnt24 = Font(app, size=24)
		self.fnt30 = Font(app, size=30)
		self.descfont = Font(app, size=app.opts['descsize'])
		self.infofont = Font(app, size=24, flags=FONT_METRICS_BASIC|FONT_METRICS_GLYPH)
	
class Vidmgr(Application):
	def handle_resolution(self):
		for (hres, vres, x, y) in self.resolutions:
			if (hres == 1280):
				return (hres, vres, x, y)
			
		self.active = False
		self.sound('bonk')
		return self.resolutions[0]
	
	def startup(self):
		self.vwInfo = None
		config = self.context.server.config
		self.opts = Config.load(config)
					
		# get the tivo information out of the startup config file.  For each tivo, we need to know:
		# tivox.name - the user friendly name and
		# tivox.tsn - the TSN
		# these fields all go into a section named [tivos]		
		self.loadTivos(config)
		if len(self.tivos) == 0:
			raise ConfigError("No Tivos found - exiting")

		# get the pytivo information.  For each pytivo instance, we need the following:
		# pytivox.config - the location of the config file
		# pytivox.ip - the ip address
		#
		# also, if the pytivo port number is not specified in the pytivo config file, you must have
		# pytivox.port - the port number
		self.shares = []		
		self.loadShares(config)
		if len(self.shares) == 0:
			raise ConfigError("No shares found - exiting")
		
	def cleanup(self):
		tc.save()
		if self.vcChanged:
			print "Video Cache has changed - saving"
			self.vc.save()
		
	def handle_active(self):
		self.myimages = Images(self)
		self.myfonts = Fonts(self)
		self.tdcount = 0

		self.push = None
		self.msgbox = None
		self.rebuildingCache = False
		
		self.vc = VideoCache(self.opts, self.context.server.config)
		
		self.root.set_resource(self.myimages.Background)
		self.TitleView = View(self, height=30, width=screenWidth, ypos=titleYPos)
		self.SubTitleView= View(self, height=20, width=screenWidth, ypos=subTitleYPos)
		
		self.ldm = ListDisplayManager(self, self.opts)
		self.ddm = DetailDisplayManager(self, self.opts, tc)
		self.vwInfo = InfoView(self, self.opts, self.myfonts.infofont)
		self.subMenu = SubMenu(self, self.opts)

		self.rootNode = self.vc.load()
		self.vcChanged = False

		self.start()

	def reStart(self):
		self.ldm.ReInit()
		self.start()
		
	def start(self):
		if self.rootNode == None:
			raise ConfigError("No video cache - exiting")

		self.TitleView.set_text(TITLE, font=self.myfonts.fnt30, colornum=0xffffff, flags=RSRC_VALIGN_BOTTOM)
		self.SubTitleView.set_text(self.rootNode.getFullTitle(), font=self.app.myfonts.fnt20, colornum=0xffffff, flags=RSRC_VALIGN_BOTTOM)

		self.currentNode, self.currentItem = self.ldm.Descend(self.rootNode)
		self.lopts = self.rootNode.getOpts()
		self.ddm.show(self.currentItem)
		
		if self.currentNode == None:
			self.active = False
			
	def handle_font_info(self, font):
		if self.vwInfo == None:
			print "Got handle font before vwInfo was instantiated ??"
		else:
			self.vwInfo.setinfogeometry(font)
			
	def handle_key_press(self, keynum, rawcode):
		if keynum == KEY_THUMBSDOWN:
			self.tdcount += 1
			if self.tdcount == 3:
				self.mb = MessageBox(self.app, "Rebuilding Cache...",
					"Please Wait",
					'alert',
					[[[KEY_TIVO], 0]])

				self.send_key(KEY_TIVO, MYKEY_REBUILDCACHE)
				self.rebuildingCache = True
				return

			self.sound('thumbsup')
			return
		else:
			self.tdcount = 0
			
		if keynum == KEY_TIVO and rawcode == MYKEY_REBUILDCACHE:
			self.rootNode = self.vc.build()
			self.vcChanged = True

			self.mb.close()
			self.reStart()
			return

		if self.msgbox:
			self.msgbox.handlekeypress(keynum, rawcode)
			
		elif self.vwInfo.isVisible:
			snd = 'updown'
			if keynum in [ KEY_LEFT, KEY_CLEAR, KEY_INFO ]:
				self.vwInfo.hide()
				
			elif keynum in [ KEY_DOWN, KEY_CHANNELDOWN ]:
				if not self.vwInfo.pagedown():
					snd = 'bonk'
					
			elif keynum in [ KEY_UP, KEY_CHANNELUP ]:
				if not self.vwInfo.pageup():
					snd = 'bonk'
					
			else:
				snd = 'bonk'
				
			self.sound(snd)

		elif self.push:
			if keynum == KEY_TIVO and rawcode == MYKEY_PUSHCOMPLETE:
				self.push = None
			else:
				self.push.handlekeypress(keynum, rawcode)
				
		elif self.currentItem == None and keynum not in [ KEY_LEFT, KEY_TIVO ]:
			self.sound('bonk')
				
		elif keynum == KEY_INFO:
			meta = self.currentItem.getMeta()
			if len(meta) == 0:
				snd = 'bonk'
			else:
				self.vwInfo.loadmetadata(meta)
				self.vwInfo.paint()
				self.vwInfo.show()
				snd = 'updown'
			self.sound(snd)
				
		elif self.ldm.isNavKey(keynum, rawcode):
			self.currentItem = self.ldm.Navigate(keynum, rawcode)
			self.ddm.show(self.currentItem)
			
		elif keynum == KEY_LEFT:
			self.currentNode, self.currentItem = self.ldm.Ascend()
			if self.currentNode == None:
				self.active = False
				return
			self.SubTitleView.set_text(self.currentNode.getFullTitle(),
						font=self.app.myfonts.fnt20,
						colornum=0xffffff, flags=RSRC_VALIGN_BOTTOM)
			self.ddm.show(self.currentItem)
			self.sound('updown')
			
		elif keynum in [KEY_RIGHT, KEY_SELECT]:
			if isinstance(self.currentItem, VideoFile):
				self.push = Push(self, self.opts, self.currentItem)
			else:
				self.SubTitleView.set_text(self.currentItem.getFullTitle(),
								font=self.app.myfonts.fnt20,
								colornum=0xffffff, flags=RSRC_VALIGN_BOTTOM)
				self.lopts = self.currentItem.getOpts()
				self.currentNode, self.currentItem = self.ldm.Descend(self.currentItem)
			self.ddm.show(self.currentItem)
			self.sound('updown')
			
		elif keynum == KEY_CLEAR:
			if not self.deletable(self.currentItem):
				self.sound('bonk')
			else:
				MessageBox(self.app, "Delete Confirmation",
						"Press THUMBS-UP button to confirm deletion of " + 
						self.currentItem.formatDisplayText(self.lopts['dispopt']),
						'alert',
						[
						 [[KEY_THUMBSUP], MYKEY_DELETECONFIRM], 
						 [[], MYKEY_DELETECANCEL]
						])
				
		elif keynum == KEY_TIVO and rawcode == MYKEY_DELETECOMPLETE:
			pass # nothing to do here
		
		elif keynum == KEY_TIVO and rawcode == MYKEY_DELETECANCEL:
			m = MessageBox(self.app, "Delete Cancelled",
					"CANCELLED",
					'bonk',
					[[[KEY_TIVO], 0]])
			self.sleep(1)
			m.close()
		
		elif keynum == KEY_TIVO and rawcode == MYKEY_DELETECONFIRM:
			m = MessageBox(self.app, "Deleting...",
					"Please Wait",
					'updown',
					[[[KEY_TIVO], 0]])
			self.currentItem.delVideo()
			self.currentItem = self.ldm.Verify()
			self.ddm.show(self.currentItem)
			m.close()
			MessageBox(self.app, "Deleted",
					"Video/Metadata/Artwork Deleted - Press any key",
					'alert',
					[[[], MYKEY_DELETECOMPLETE]])
			self.vcChanged = True
		
		else:
			self.sound('bonk')

	def deletable(self, item):
		if not isinstance(item, VideoFile):
			return False
		
		if item.isDeletable():
			return True
		
		return False
		
	# load up tivo information from the config file
	def loadTivos(self, cfg):
		def cmptivo (left, right):
			return cmp(left['name'], right['name'])
		
		tlist = []
		section = 'tivos'
		
		allchars = maketrans('', '')
		if cfg.has_section(section):
			i = 0
			while (True):
				i = i + 1
				namekey = 'tivo' + str(i) + '.name'
				tsnkey = 'tivo' + str(i) +  '.tsn'
				if cfg.has_option(section, namekey) and cfg.has_option(section, tsnkey):
					tlist.append({'name' : cfg.get(section, namekey),
									'tsn' : cfg.get(section, tsnkey).translate(allchars, '-')})
				else:
					break
				
		self.tivos = sorted(tlist, cmp=cmptivo)

	# load up pytivo and shares information from config and from pytivo config(s)
	def loadShares(self, cfg):
		self.shares = []
		
		s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
		s.connect(('4.2.2.1', 123))
		defip = s.getsockname()[0]
	
		section = 'pytivos'
		if cfg.has_section(section):
			i = 0
			while (True):
				i = i + 1
				key = "pytivo" + str(i) + ".config"
				if not cfg.has_option(section, key): break
				cfgfile = cfg.get(section, key)
				
				sep = None
				sepkey = 'pytivo' + str(i) + '.sep'
				if cfg.has_option(section, sepkey): sep = cfg.get(section, sepkey)
				
				ip = defip
				key = "pytivo" + str(i) + ".ip"
				if cfg.has_option(section, key):
					ip = cfg.get(section, key)

				port = None				
				key = "pytivo" + str(i) + ".port"
				if cfg.has_option(section, key):
					port = cfg.get(section, key)
					
				key = "pytivo" + str(i) + ".skip"
				skip = []
				if cfg.has_option(section, key):
					sk = cfg.get(section, key).split(",")
					skip = [s.strip() for s in sk]
				
				self.loadPyTivoConfig(cfgfile, ip, port, sep, skip)
					
				if port == None:
					raise ConfigError("Neither main config file nor pytivo config file " + cfgfile + " has port number specified")

	# parse a pytivo config looking for shares				
	def loadPyTivoConfig(self, cf, ip, defport, sep, skip):
		pyconfig = ConfigParser.ConfigParser()
		if not pyconfig.read(cf):
			raise ConfigError("ERROR: pyTivo config file " + cf + " does not exist.")

		port = defport
		if pyconfig.has_option('Server', 'port') : port = pyconfig.get('Server', 'port')
		
		for section in pyconfig.sections():
			if not section in skip:
				if (pyconfig.has_option(section, "type")
						and (pyconfig.get(section, "type") == "video" or pyconfig.get(section, "type") == "dvdvideo")
						and	pyconfig.has_option(section, 'path')):
					path = pyconfig.get(section, 'path')
					self.shares.append({'name' : section,
							'ip' : ip,
							'port' : port,
							'path' : path,
							'sep' : sep})

		
